Add labels to explicitly set block type

gpetiot opened the pull request:
Fix #164 
This adds an intermediary step checking what is guessed vs what is enforced with this new label. Some blocks (ocaml, cram, toplevel) can be forced into a raw block (when it may be invalid on purpose, but the markdown header is set for highlighting)

NathanReb commented on lib/block.ml:
```diff
@@ -365,28 +368,42 @@ let mk ~line ~file ~column ~section ~labels ~legacy_labels ~header ~contents
           check_no_errors errors >>= fun () ->
           check_not_set "`env` label cannot be used with a `shell` header." env
-          >>= fun () -> Ok (Cram { non_det })
+          >>= fun () -> Ok (`Cram { non_det })
```
Any particular reason why we're using polymorphic variants again? If the name collide I think we should just scope them in a submodule or use different names for the kind alone and for the block themselves.

What do you think?

gpetiot replied:
Fixed.

NathanReb commented on lib/label.ml:
```diff
@@ -74,6 +74,8 @@ type non_det = Nd_output | Nd_command
 
 let default_non_det = Nd_output
 
+type block_kind = Raw | OCaml | Cram | Toplevel | Include
```
My impression is that Raw is unecessary as it has no value to mdx, it simply won't be interpreted. I think the skip label already covers that case.

gpetiot replied:
I didn't think of this, if using `skip` fixes #164 I guess this PR does not have any purpose for now until we make the block type label mandatory

NathanReb commented on lib/block.ml:
```diff
@@ -365,28 +368,42 @@ let mk ~line ~file ~column ~section ~labels ~legacy_labels ~header ~contents
+      | _ -> check_no_errors errors >>= fun () -> Ok (`Raw { header }) ) )
+  >>= fun guess ->
+  ( match (guess, block_kind) with
+  | `Include b, (None | Some Include) -> Ok (Include b)
```
I'm not sure that's how it should work. I think the explicit block kind should prevail no matter what.

If no block kind is specified, then we try to infer what it is based on the header, the labels, the syntax etc.

One of the main reason why those are useful is that they allow for clear error reporting and easy to document behaviour.

gpetiot replied:
I wrote it this way because we have to check the block labels whether the block kind is specified or not. Let's say I set the block kind to `toplevel` I have to check all the labels corresponding to the record fields are set (to fill the record), and that none of the incompatible labels are set. That's why I check the labels first, and then check that what is guessed from them is consistent with the block kind if it exists.

NathanReb replied:
I haven't taken a deep enough look but that seems dubious to me. I feel like the `guess` is already the result of the inference. That means that the error will always be the one from inference. Also you if the inferred block differs from the explicit one, the error is always the same. I feel like that's not really the intended behaviour.

Could you add tests to make sure this behaves properly?

gpetiot replied:
I redesigned the whole code, I still have to add some tests.

gpetiot commented:
> E.g. if you set the block to toplevel but use regular ocaml syntax (no leading `#` and proper line continuation) you get an approriate error rather than a silent change of behaviour.

I rewrote the whole PR and added more precise error messages (and some tests about this particular toplevel/ocaml error)

gpetiot commented on test/bin/mdx-test/expect/empty/test-case.md:
````diff
@@ -7,7 +7,7 @@ Testing empty blocks
 
 <!-- $MDX env=x -->
-```foo
+```ocaml
````
To avoid having a `env is not allowed for raw blocks` message, which is not the purpose of the test.

NathanReb commented on lib/block.ml:
```diff
@@ -295,105 +296,195 @@ let version_enabled version =
+  | Some Toplevel -> mk_toplevel ~config ~header ~contents ~errors
+  | Some Include -> mk_include ~config ~header ~errors
+  | None ->
+      mk_ocaml ~config ~header ~contents ~errors
```
I think this should fallback to the old block type inference we used to have. That should ensure we don't break the old behaviour + will perform better and allow better error messages than sequentially trying to interpret it as any of the block types.

In particular if we change the order here it will change the error message and that's highly confusing to users.

gpetiot replied:
Agreed, done.

NathanReb commented on lib/block.ml:
```diff
@@ -295,105 +296,195 @@ let version_enabled version =
+
+let fmt_header = Fmt.using (Format.asprintf "`%a`" Header.pp) Fmt.string
+
+let header_not_allowed ~header ~kind =
```
I wonder if we should really forbid any headers. When the block type is set explicitly, the header shouldn't matter. When the block type isn't set explicitly, we use it to determine the block type so this will never happen.

What do you think?

gpetiot replied:
I think it's better for consistency to forbid headers incompatible for blocks of type T, when said headers are incompatible with T. This way, when we remove the block type constraint the behavior of the block would be the same.

NathanReb replied:
The block header doesn't have an actual semantic value, it's just used for styling and in our case, to infer the block type when that's the only way. They're not incompatible, they just mess with our inference logic. If we set their type explicitly, mdx shouldn't try to reason about them as they are irrelevant to mdx. I feel like it's a bit intrusive.

Imagine I want to apply a custom style to some OCaml blocks and to do so I set the block header to `custom-ocaml` and explicitly set the mdx block type to ocaml. With the current behaviour, mdx will fail to parse the block, telling me I can't use this header for this block type. Even though this scenario is a bit of a fiction and I don't of any such cases, I think it's possible as some md interpreters just use the block header to set a class in the generated html div.
In such a case it would be incredibly frustrating to have it refused by mdx for no substantial reason since the block header prevents in no way mdx from interpreting the block.

gpetiot replied:
I see, good point.

gpetiot replied:
Agreed, it makes more sense, I will simplify this part of the code.

NathanReb commented on lib/block.ml:
```diff
@@ -295,105 +296,195 @@ let version_enabled version =
-      check_not_set "`env` label cannot be used with a `file` label." env
-      >>= fun () ->
-      check_no_errors errors >>= fun () ->
+type label_config = {
```
I think using a record here makes much more sense than what we use to have but `label_config` is a bit confusing. How about a simpler `labels` or eventually `block_config`?

NathanReb commented on lib/block.ml:
```diff
@@ -304,45 +311,108 @@ let check_no_errors = function
+let infer_block ~config ~header ~contents ~errors =
+  match config with
+  | { file_inc = Some file_included; non_det; env; part; _ } -> (
       check_not_set
```
would it make sense to use `mk_include` here or is the error reporting different?

NathanReb replied:
Same remark for the other cases, i.e. once we have infered the type of block, can't use the corresponding `mk_<block-type>` function?
